File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/activity/ActivityScopedFactoryBeanHandlerTest.java
package rocks.frieler.android.beans.scopes.activity;

import android.app.Activity;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;

import androidx.fragment.app.FragmentActivity;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.clearInvocations;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ActivityScopedFactoryBeanHandlerTest {
    private ForegroundActivityHolder foregroundActivityHolder = mock(ForegroundActivityHolder.class);
    private ActivityScopedFactoryBeanHandler activityScopedFactoryBeanHandler = new ActivityScopedFactoryBeanHandler(foregroundActivityHolder);

    @Test
    public void testGetNameReturnsActivityScopeName() {
        assertThat(activityScopedFactoryBeanHandler.getName(), is(ActivityScopedFactoryBeanHandler.ACTIVITY_SCOPE));
    }

    @Test
    public void testScopeIsNotActiveWhenNoActivityIsInTheForeground() {
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(null);

        final boolean active = activityScopedFactoryBeanHandler.isActive();

        assertThat(active, is(false));
    }

    @Test
    public void testScopeIsNotActiveWhenTheActivityInTheForegroundIsNotAFragmentActivity() {
        final Activity activity = mock(Activity.class);
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(activity);

        final boolean active = activityScopedFactoryBeanHandler.isActive();

        assertThat(active, is(false));
    }

    @Test
    public void testScopeIsActiveWhenAFragmentActivityIsInTheForeground() {
        final Activity fragmentActivity = mock(FragmentActivity.class);
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(fragmentActivity);

        final boolean active = activityScopedFactoryBeanHandler.isActive();

        assertThat(active, is(true));
    }

    @Test
    public void testGetBeanReturnsBeanProducedByFactoryBean() {
        final Activity fragmentActivity = Robolectric.buildActivity(FragmentActivity.class).get();
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(fragmentActivity);

        @SuppressWarnings("unchecked")
        final ActivityScopedFactoryBean<ActivityScopedFactoryBeanHandlerTest> activityScopedFactoryBean = mock(ActivityScopedFactoryBean.class);
        when((activityScopedFactoryBean).produceBean()).thenReturn(this);

        final ActivityScopedFactoryBeanHandlerTest bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean);

        assertThat(bean, is(sameInstance(this)));
    }

    @Test
    public void testGetBeanReturnsBeanAlreadyPresentInActivityScope() {
        final Activity fragmentActivity = Robolectric.buildActivity(FragmentActivity.class).get();
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(fragmentActivity);

        @SuppressWarnings("unchecked")
        final ActivityScopedFactoryBean<ActivityScopedFactoryBeanHandlerTest> activityScopedFactoryBean = mock(ActivityScopedFactoryBean.class);
        when((activityScopedFactoryBean).produceBean()).thenReturn(this);

        final ActivityScopedFactoryBeanHandlerTest bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean);
        verify(activityScopedFactoryBean).produceBean();
        assertThat(bean, is(sameInstance(this)));
        //noinspection unchecked
        clearInvocations(activityScopedFactoryBean);

        final ActivityScopedFactoryBeanHandlerTest beanAgain = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean);
        verify(activityScopedFactoryBean, never()).produceBean();
        assertThat(beanAgain, is(sameInstance(bean)));
    }

    @Test
    public void testGetBeanSetsActivityOnActivityAwareBean() {
        final Activity fragmentActivity = Robolectric.buildActivity(FragmentActivity.class).get();
        when(foregroundActivityHolder.getCurrentActivity()).thenReturn(fragmentActivity);

        final ActivityAware activityAwareBean = mock(ActivityAware.class);
        @SuppressWarnings("unchecked")
        final ActivityScopedFactoryBean<ActivityAware> activityScopedFactoryBean = mock(ActivityScopedFactoryBean.class);
        when(activityScopedFactoryBean.produceBean()).thenReturn(activityAwareBean);

        final ActivityAware bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean);

        verify(activityAwareBean).setActivity(fragmentActivity);
        assertThat(bean, is(sameInstance(activityAwareBean)));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/activity/ActivityScopedFactoryBeanHandlerTest.kt
package rocks.frieler.android.beans.scopes.activity

import android.app.Activity
import androidx.activity.ComponentActivity
import assertk.assertThat
import assertk.assertions.*
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.clearInvocations
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import rocks.frieler.android.beans.BeansProvider

@RunWith(RobolectricTestRunner::class)
@Config(manifest=Config.NONE)
class ActivityScopedFactoryBeanHandlerTest {
	private val foregroundActivityHolder : ForegroundActivityHolder = mock()
	private val activityScopedFactoryBeanHandler = ActivityScopedFactoryBeanHandler(foregroundActivityHolder)

	@Test
	fun `name is Activity-scope`() {
		assertThat(activityScopedFactoryBeanHandler.name).isEqualTo(ActivityScopedFactoryBeanHandler.ACTIVITY_SCOPE)
	}

	@Test
	fun `ActivityScope is not active when no Activity is in the foreground`() {
		whenever(foregroundActivityHolder.currentActivity).thenReturn(null)

		val active = activityScopedFactoryBeanHandler.isActive

		assertThat(active).isFalse()
	}

	@Test
	fun `ActivityScope is not active when the Activity in the foreground is not a FragmentActivity`() {
		val activity: Activity = mock()
		whenever(foregroundActivityHolder.currentActivity).thenReturn(activity)

		val active = activityScopedFactoryBeanHandler.isActive

		assertThat(active).isFalse()
	}

	@Test
	fun `ActivityScope is active when a ComponentActivity is in the foreground`() {
		val componentActivity: ComponentActivity = mock()
		whenever(foregroundActivityHolder.currentActivity).thenReturn(componentActivity)

		val active = activityScopedFactoryBeanHandler.isActive

		assertThat(active).isTrue()
	}

	@Test
	fun `getBean() returns bean produced by ActivityScopedFactoryBean`() {
		val dependencies: BeansProvider = mock()
		val componentActivity: Activity = Robolectric.buildActivity(ComponentActivity::class.java).get()
		whenever(foregroundActivityHolder.currentActivity).thenReturn(componentActivity)
		val activityScopedFactoryBean: ActivityScopedFactoryBean<ActivityScopedFactoryBeanHandlerTest> = mock()
		whenever(activityScopedFactoryBean.produceBean(dependencies)).thenReturn(this)

		val bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean, dependencies)

		assertThat(bean).isSameInstanceAs(this)
	}

	@Test
	fun `getBean() returns bean already present in Activity-scope`() {
		val dependencies: BeansProvider = mock()
		val componentActivity: Activity = Robolectric.buildActivity(ComponentActivity::class.java).get()
		whenever(foregroundActivityHolder.currentActivity).thenReturn(componentActivity)
		val activityScopedFactoryBean: ActivityScopedFactoryBean<ActivityScopedFactoryBeanHandlerTest> = mock()
		whenever(activityScopedFactoryBean.produceBean(dependencies)).thenReturn(this)

		val bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean, dependencies)

		verify(activityScopedFactoryBean).produceBean(dependencies)
		assertThat(bean).isSameInstanceAs(this)
		clearInvocations(activityScopedFactoryBean)

		val beanAgain = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean, dependencies)

		verify(activityScopedFactoryBean, never()).produceBean(dependencies)
		assertThat(beanAgain).isSameInstanceAs(bean)
	}

	@Test
	fun `getBean() sets Activity on ActivityAware bean`() {
		val dependencies: BeansProvider = mock()
		val componentActivity: Activity = Robolectric.buildActivity(ComponentActivity::class.java).get()
		whenever(foregroundActivityHolder.currentActivity).thenReturn(componentActivity)
		val activityAwareBean: ActivityAware = mock()
		val activityScopedFactoryBean: ActivityScopedFactoryBean<ActivityAware> = mock()
		whenever(activityScopedFactoryBean.produceBean(dependencies)).thenReturn(activityAwareBean)

		val bean = activityScopedFactoryBeanHandler.getBean("bean", activityScopedFactoryBean, dependencies)

		verify(activityAwareBean).setActivity(componentActivity)
		assertThat(bean).isSameInstanceAs(activityAwareBean)
	}
}
